File: src/main/java/com/oycm/algorithm/ba/opposite/Solution_8.scala
package com.oycm.algorithm.ba.opposite

import sun.nio.cs.ext.MacHebrew

object Solution_8 {

  /**
   * 658. 找到 K 个最接近的元素
   * https://leetcode.cn/problems/find-k-closest-elements/description/
   * 从数组中找到最靠近 x 的 k 个数
   *
   * 整数 a 比整数 b 更接近 x 需要满足：
   *  - |a - x| < |b - x|
   *  - |a - x| == |b - x| 且 a < b
   *
   * 返回的结果必须要是按升序排好的
   *
   * @param arr 升序排序 的数组
   * @param k   整数 [1, n]
   * @param x   整数 [-10000, 10000]
   * @return
   */
  def findClosestElements(arr: Array[Int], k: Int, x: Int): List[Int] = {
    /*
    把 x 看成数轴中的一点
    arr(i) 的值也是数轴中的点，将 l = 0，r = n-1 和 x 进行比较，l 与 r 到 x 的距离
    x <= l 或 x >= r 就是从边界开始循环
    x 在 arr 中，可以找出最远点，从 [0, n-1] 从后面截取 [n-k, n-1] 再排序
    */
    val n = arr.length
    val ans = Array.fill(k)(0)
    if (x <= arr(0)) {
      for (i <- ans.indices) {
        ans(i) = arr(i)
      }
    } else if (x >= arr(n - 1)) {
      for (i <- ans.indices) {
        ans(i) = arr(n - k + i)
      }
    } else {
      var l = 0
      var r = n - 1
      var i = 0
      while (l <= r) {
        val a = Math.abs(arr(l) - x)
        val b = Math.abs(arr(r) - x)
        if (a > b) {
          if (i >= n - k) {
            ans((i + k) % n) = arr(l)
          }
          l += 1
        } else if (a <= b) {
          if (i >= n - k) {
            ans((i + k) % n) = arr(r)
          }
          r -= 1
        }
        i += 1
      }
      return ans.toList.sorted
    }
    ans.toList
  }

  def main(args: Array[String]): Unit = {
    println(findClosestElements(Array(1, 2, 3, 4, 5), 4, 3))
    println(findClosestElements(Array(1, 1, 2, 3, 4, 5), 4, -1))
    println(findClosestElements(Array(0, 1, 2, 2, 2, 3, 6, 8, 8, 9), 5, 9))
  }
}

File: src/main/java/com/oycm/algorithm/i/subset/Solution_8.java
package com.oycm.algorithm.i.subset;

import java.util.List;

public class Solution_8 {

    /**
     * 1239. <a href="https://leetcode.cn/problems/maximum-length-of-a-concatenated-string-with-unique-characters/description/">串联字符串的最大长度</a> 1720
     *
     * @param arr 字符串数组
     * @return 返回所有可行解 s 中最长长度
     */
    public int maxLength(List<String> arr) {
        /*
        字符串 s 是将 arr 的含有 不同字母 的 子序列 字符串 连接 所得的字符串
         */
        /*
        不能有相同字母
        // todo 可位运算优化
         */
        dfs(0, arr, new char[26], 0);
        return ans;
    }

    int ans = 0;

    public void dfs(int i, List<String> arr, char[] cs, int length) {
        if (i == arr.size()) {
            ans = Math.max(ans, length);
            return;
        }
        dfs(i + 1, arr, cs, length);
        /*
        判断是否能选 arr[i]
         */
        boolean enable = true;
        for (char c : arr.get(i).toCharArray()) {
            c -= 'a';
            cs[c]++;
            if (cs[c] > 1) {
                enable = false;
            }
        }
        if (enable) {
            dfs(i + 1, arr, cs, length + arr.get(i).length());
        }
        // 回溯
        for (char c : arr.get(i).toCharArray()) {
            c -= 'a';
            cs[c]--;
        }
    }


}
